Refactor fips jammy

ramonskie opened the pull request:
cherry picked pr's that where meant for bionic in to jammy
#229 
#236
#237 

this should give us a starting point for jammy fips stemcells.

rkoster commented:
@toabctl do you see any problems with merging this PR? Knowing that this will not give us a functional stemcell until the packages are available for jammy, this should at least unblock the changes needed to stemcell pipeline to produce private FIPS variants (as discussed in https://github.com/cloudfoundry/community/pull/705).

toabctl commented:
@rthill91 could you have a look please?

rthill91 commented:
This is not an exhaustive review, but a couple things that stood out on first skim:

- I see a lot of changes made to the `ubuntu_bionic_spec`s that should probably be targeting `ubuntu_jammy_spec` for this PR
- Duplication between `base_apt_fips`, `base_fips_apt`, and `prelude_fips`
  - To keep inline with the `ubuntu-bionic/master` branch I'd recommend putting all the fips logic from those 3 files in `prelude_fips`, sourcing that in `base_fips_apt` and dropping `base_apt_fips` entirely

ramonskie commented:
@rthill91 i cherry picked all the fips commits from the bionic branch. it seems i missed 1
now i update the repo with your suggestions.

as for the spec files. as the jammy fips packges are not updated yet. i thought it would best to leave it for now

ramonskie commented:
okay i see what happend i cherrypicked in the wrong order from new to old. i will redo this

rthill91 commented on stemcell_builder/stages/base_ubuntu_packages/apply.sh:
```diff
@@ -85,5 +86,10 @@ run_in_chroot "${chroot}" "systemctl disable systemd-resolved"
 
+run_in_chroot "${chroot}" "systemctl disable chrony"
+
+pkgs_to_purge="crda iw mg wireless-crda wireless-regdb"
```
`crda` and `wireless-crda` do not exist for jammy, I think this will break both fips and nonvariant builds

rthill91 replied:
Removing those from the purge list I see several spec failures introduced. I haven't had a chance to look at them in depth at all, but to save someone else some time, here's the list of failures
```sh
Finished in 11.24 seconds (files took 0.10947 seconds to load)                                                                                                                                                                                                                                                                                 
407 examples, 15 failures                                                                                                                                                                                                                                                                                                                      
                                                                                                                                                                                                                                                                                                                                               
Failed examples:                                                                                                                                                       
                                                                                                                                                                       
rspec './spec/os_image/ubuntu_jammy_spec.rb[1:3:1:3:1]' # Ubuntu 22.04 OS image behaves like a systemd-based OS image systemd services chrony is expected to be enabled
rspec './spec/os_image/ubuntu_jammy_spec.rb[1:6:1:1:2:1]' # Ubuntu 22.04 OS image behaves like a Linux kernel module configured OS image prevent bluetooth module from being loaded (stig: V-38682) /etc/modprobe.d/blacklist.conf content is expected to match "install bluetooth /bin/true"
rspec './spec/os_image/ubuntu_jammy_spec.rb[1:6:2:1:1:1]' # Ubuntu 22.04 OS image behaves like a Linux kernel module configured OS image prevent tipc module from being loaded (stig: V-38517) /etc/modprobe.d/blacklist.conf content is expected to match "install tipc /bin/true"
rspec './spec/os_image/ubuntu_jammy_spec.rb[1:6:3:1:1:1]' # Ubuntu 22.04 OS image behaves like a Linux kernel module configured OS image prevent sctp module from being loaded (stig: V-38515) /etc/modprobe.d/blacklist.conf content is expected to match "install sctp /bin/true"          
rspec './spec/os_image/ubuntu_jammy_spec.rb[1:6:4:1:1:1]' # Ubuntu 22.04 OS image behaves like a Linux kernel module configured OS image prevent dccp module from being loaded (stig: V-38514) /etc/modprobe.d/blacklist.conf content is expected to match "install dccp /bin/true"
rspec './spec/os_image/ubuntu_jammy_spec.rb[1:6:6:1:1:1]' # Ubuntu 22.04 OS image behaves like a Linux kernel module configured OS image prevent USB module from being loaded (stig: V-38490) /etc/modprobe.d/blacklist.conf content is expected to match "install usb-storage /bin/true"
rspec './spec/os_image/ubuntu_jammy_spec.rb[1:6:7:1:1:1]' # Ubuntu 22.04 OS image behaves like a Linux kernel module configured OS image prevent ipv6 module from being loaded (stig: V-38546) (stig: V-38444) (stig: V-38553) (stig: V-38551) /etc/modprobe.d/blacklist.conf content is expected to match "options ipv6 disable=1"
rspec './spec/os_image/ubuntu_jammy_spec.rb[1:6:8:1:1:1]' # Ubuntu 22.04 OS image behaves like a Linux kernel module configured OS image prevent cramfs module from being loaded (CIS-2.18) /etc/modprobe.d/blacklist.conf content is expected to match "install cramfs /bin/true"       
rspec './spec/os_image/ubuntu_jammy_spec.rb[1:6:9:1:1:1]' # Ubuntu 22.04 OS image behaves like a Linux kernel module configured OS image prevent freevxfs module from being loaded (CIS-2.19) /etc/modprobe.d/blacklist.conf content is expected to match "install freevxfs /bin/true"                                             
rspec './spec/os_image/ubuntu_jammy_spec.rb[1:6:10:1:1:1]' # Ubuntu 22.04 OS image behaves like a Linux kernel module configured OS image prevent jffs2 module from being loaded (CIS-2.20) /etc/modprobe.d/blacklist.conf content is expected to match "install jffs2 /bin/true" 
rspec './spec/os_image/ubuntu_jammy_spec.rb[1:6:11:1:1:1]' # Ubuntu 22.04 OS image behaves like a Linux kernel module configured OS image prevent hfs module from being loaded (CIS-2.21) /etc/modprobe.d/blacklist.conf content is expected to match "install hfs /bin/true"         
rspec './spec/os_image/ubuntu_jammy_spec.rb[1:6:12:1:1:1]' # Ubuntu 22.04 OS image behaves like a Linux kernel module configured OS image prevent hfsplus module from being loaded (CIS-2.22) /etc/modprobe.d/blacklist.conf content is expected to match "install hfsplus /bin/true"
rspec './spec/os_image/ubuntu_jammy_spec.rb[1:6:13:1:1:1]' # Ubuntu 22.04 OS image behaves like a Linux kernel module configured OS image prevent squashfs module from being loaded (CIS-2.23) /etc/modprobe.d/blacklist.conf content is expected to match "install squashfs /bin/true"
rspec './spec/os_image/ubuntu_jammy_spec.rb[1:6:14:1:1:1]' # Ubuntu 22.04 OS image behaves like a Linux kernel module configured OS image disable RDS (CIS-7.5.3) /etc/modprobe.d/blacklist.conf content is expected to match "install rds /bin/true"                                
rspec ./spec/os_image/ubuntu_jammy_spec.rb:41 # Ubuntu 22.04 OS image installed by system_kernel linux-generic-hwe-22.04 is expected to be installed
```

ramonskie replied:
im trying to debug these issues now. i don't fully understand where this is coming from at the moment
so i converted this pr to a draft

ramonskie replied:
so i found the issue. it was due to the fact that function 'kernel_stages` is never called.
this issue is also persistent in the `ubuntu-bionic/master` branch

example:
this function https://github.com/cloudfoundry/bosh-linux-stemcell-builder/blob/f8d91b2eb3dff351e31a6257edf67dba052ea345/bosh-stemcell/lib/bosh/stemcell/stage_collection.rb#L17
should be called on line
https://github.com/cloudfoundry/bosh-linux-stemcell-builder/blob/f8d91b2eb3dff351e31a6257edf67dba052ea345/bosh-stemcell/lib/bosh/stemcell/stage_collection.rb#L264

ramonskie committed:
add the ability to create jammy fips stemcells.
based on the commits for bionic on the following prs #210 #229 #236 #237

ramonskie commented:
i refactored this pr a bit. and builds correctly now. the only thing i cannot test completely the fips jammy part as the jammy-fips packages is not available yet

rthill91 commented:
@ramonskie I think this is good to go now on unblocking builds. I expect some smaller changes to be needed when the fips packages are available (kernel version, spec changes for fips, etc), but that's a problem for later

aramprice commented on stemcell_builder/lib/prelude_fips.bash:
```diff
@@ -0,0 +1,141 @@
+    echo "FIPS variant detected. setting up FIPS"
+fi
+
+if [ -z ${UBUNTU_ADVANTAGE_TOKEN} ]; then
+    echo "'fips' variant detected but \$UBUNTU_ADVANTAGE_TOKEN not given."
```
It seems like we are conflating `UBUNTU_ADVANTAGE_TOKEN` and "FIPS variant" here. The existence of `UBUNTU_ADVANTAGE_TOKEN` is a (possibly required) implementation detail of assembling a FIPS variant. 

I would suggest that we use an explicit "FIPS" variable here since an `UBUNTU_ADVANTAGE_TOKEN` could also be used in cases when a stemcell is being built with Extended Support packages.

ramonskie committed:
do not put variant type in the operating_system in the manifest.mf of the stemcell

ramonskie committed:
fix unit tests after all the fips cherrypicking

aramprice commented on bosh-stemcell/lib/bosh/stemcell/stage_collection.rb:
```diff
@@ -16,9 +16,16 @@ def operating_system_stages
 
     def kernel_stages
       if operating_system.variant == "fips"
```
Could `"fips"` be pulled out as a constant somewhere?

aramprice commented on bosh-stemcell/lib/bosh/stemcell/stage_collection.rb:
```diff
@@ -14,6 +14,21 @@ def operating_system_stages
+        %I[
+          system_#{infrastructure.name}_fips_kernel
+          base_fips_apt
+          system_kernel_modules
```
Should `system_kernel_modules` be pulled out from like 27

ramonskie replied:
yes i just fixed this yesterday.

ramonskie committed:
use the kernel_stages in stemcell phase instead of the os phase

rkoster reviewed (changes requested):
@ramonskie could you put: https://github.com/cloudfoundry/bosh-linux-stemcell-builder/pull/300/commits/1e202baf85e727e4d61fd02912609fc869545636 in a separate PR?

ramonskie committed:
[WIP] this commit can be reverted when fips packages are available.
to use and verify the fips stemcell builder proccess
we need to disable some fips specific stages. so that we create a bogus fips stemcell for testing purposes
